### ISSUE TITLE
Sigmoid only applies on Mu output, not sigma output

### ISSUE THREAD

**Issue by gatoniel** [CSBDeep/CSBDeep#4]
Hi,

in #3 I explained what I did. My moisy input data and the ground truth are all normalized to the range 0...1 in numpy.float32. That is why I want to apply a _sigmoid_ as the last activation. Linear and _relu_ in the end makes way to big outputs. With my metrics I also check the maximum, minimum,, mean and median values of the Mu outputs and the Sigma outputs of a probabilistic model. When applying _sigmoid_ to the outputs I get Mu in the range of 0...1. But the Sigma output is not in that range, as you can see:

> 861/862 [============================>.] - ETA: 0s - loss: -1.0504 - mse: 0.0519 - mae: 0.1333 - laplacian_loss: -1.7435 - psnr: 36.9319 - mse_1: 0.0519 - ssim: 0.3454 - ms_ssim: 0.6584 - pearson_corr_coeff: 0.2762 - min_mu: 0.0565 - max_mu: 0.5003 - min_sigma: 0.0917 - max_sigma: 2.3240 - mean_sigma: 1.1987 - median_sigma: 1.2540 - min_sigma_mean: 0.0977 - max_sigma_mean: 2.1940 - mean_sigma_mean: 1.1987 - median_sigma_mean: 1.2706 - max_difference_gt: 0.8516 - min_difference_gt: 4.4287e-07

**laplacian_loss** is the loss function I have written applied to the model as a metric. The metrics are defined in functions_csbdeep.py in my repository.
Can you explain, why the sigma output is not in the range 0...1?

**Comment by uschmidt83** [comment]
> My moisy input data and the ground truth are all normalized to the range 0...1 in numpy.float32.

How exactly do you normalize your data? Have you tried the default percentile-based normalization?

> That is why I want to apply a sigmoid as the last activation. Linear and relu in the end makes way to big outputs.

If your data is normalized to the range 0...1, then the network should learn to restore values in that range. I don't understand why the network would make "way to big outputs".

> When applying sigmoid to the outputs I get Mu in the range of 0...1. But the Sigma output is not in that range, as you can see: ...

Sigma is the scale parameter of a Laplace distribution, which is ensured to be positive by using a *softplus* activation function in the network. Hence, using *sigmoid* for `unet_last_activation` has no effect on sigma.

In general, it would be very helpful for us to understand why the provided functionality is not sufficient for your particular application.

**Comment by gatoniel** [comment]
Thanks for your fast answer!

> How exactly do you normalize your data? Have you tried the default percentile-based normalization?

We get our noisy data and ground truth from a simulation. The simulation generates .tif-Files with the uint16 format. We normalize our noisy data with:

```
def normalize_from_uint16_data(img):
    """
    Is used to convert the noisy pictures to training data in the range of
    0...1 in float32
    """
    return (img / (2 ** 16 - 1)).astype(np.float32)
```

Our ground truth is normalized via:

```
def normalize_from_uint16_boost(img):
    """
    Is used to convert the convoluted pictures to lable data in the range of
    0...1 in float32. Since the pictures of the simulation are not in the
    range of 0...2**16-1, but in a smaller range, we extend the range manually
    to 0...1 by dividing by the max value.
    """
    return (img / np.max(img)).astype(np.float32)
```

Since we want the highest value to be 1.0. These functions are applied to the whole image, not only patches. The normalization is applied in that way, since we want to have a clear way to save our outputs to a .tif-File. By multiplying our outputs with 2**16-1 we get pictures in the range of uint16.

> If your data is normalized to the range 0...1, then the network should learn to restore values in that range. I don't understand why the network would make "way to big outputs".

We had this problem before, when using our own networks. That is why we decided to have outputs in the range of 0...1 via _sigmoid_. I don´t think the network can push the output range to 0...1 when it has a linear or relu as last activation, since 0...1 is a small subset of 0...infinity.

We try to reproduce your results with own networks. Now that you´re code is available we want to compare the results. In our networks we applied _sigmoid_ as last activation also to the sigma outputs. 

How do you translate your outputs back to images, that humans can interpret? Maybe that answers our understanding problem. Since you don´t know exactly the range of outputs, you can´t apply any easy form of translation. Are you cutting of values that are not in the sufficient range?

**Comment by uschmidt83** [comment]
> We get our noisy data and ground truth from a simulation. The simulation generates .tif-Files with the uint16 format.

It would be much easier for me to understand this if you could give me access to a (single) raw data image pair (noisy input with corresponding ground truth) straight from your simulation without any further processing.

> I don´t think the network can push the output range to 0...1 when it has a linear or relu as last activation, since 0...1 is a small subset of 0...infinity.

This should work without issues, given proper training data. All our examples use a *linear* last activation and output values (mostly) in the 0...1 range.

> How do you translate your outputs back to images, that humans can interpret?

I don't understand. The output should look much better after it has been restored by the network and should thus also make subsequent processing easier, including visualization.

Do you mean how we "undo" the normalization required for the network?

**Comment by uschmidt83** [comment]
I just sent you an email with a trained model and the CARE result for your data example.
Hope that helps.

Uwe

**Comment by uschmidt83** [comment]
I don't understand really understand your questions. If you want, we can have a phone call about it. Just send me an email.

**Comment by maweigert** [comment]
Closing due to non-activity (please reopen if still relevant)
